torch/fx/node.py

```diff
@@ -56,6 +56,40 @@
     _ops.profiler._record_function_enter_new,
     _ops.profiler._record_function_exit,
     _ops.inductor.accumulate_grad_.default,
+    # native functional collective ops with default overload name
+    _ops._c10d_functional.wait_tensor.default,
+    _ops._c10d_functional.all_gather_into_tensor_out.default,
+    _ops._c10d_functional.all_gather_into_tensor.default,
+    _ops._c10d_functional.all_gather_into_tensor_coalesced.default,
+    _ops._c10d_functional.all_reduce.default,
+    _ops._c10d_functional.all_reduce_.default,
+    _ops._c10d_functional.all_reduce_coalesced.default,
+    _ops._c10d_functional.all_reduce_coalesced_.default,
+    _ops._c10d_functional.all_to_all_single.default,
+    _ops._c10d_functional.broadcast.default,
+    _ops._c10d_functional.broadcast_.default,
+    _ops._c10d_functional.reduce_scatter_tensor.default,
+    _ops._c10d_functional.reduce_scatter_tensor_coalesced.default,
+    _ops._c10d_functional_autograd.all_to_all_single.default,
+    _ops._c10d_functional_autograd.reduce_scatter_tensor.default,
+    _ops._c10d_functional_autograd.all_gather_into_tensor.default,
+    # native functional collective ops
+    _ops._c10d_functional.wait_tensor,
+    _ops._c10d_functional.all_gather_into_tensor_out,
+    _ops._c10d_functional.all_gather_into_tensor,
+    _ops._c10d_functional.all_gather_into_tensor_coalesced,
+    _ops._c10d_functional.all_reduce,
+    _ops._c10d_functional.all_reduce_,
+    _ops._c10d_functional.all_reduce_coalesced,
+    _ops._c10d_functional.all_reduce_coalesced_,
+    _ops._c10d_functional.all_to_all_single,
+    _ops._c10d_functional.broadcast,
+    _ops._c10d_functional.broadcast_,
+    _ops._c10d_functional.reduce_scatter_tensor,
+    _ops._c10d_functional.reduce_scatter_tensor_coalesced,
+    _ops._c10d_functional_autograd.all_to_all_single,
+    _ops._c10d_functional_autograd.reduce_scatter_tensor,
+    _ops._c10d_functional_autograd.all_gather_into_tensor,
```

Comment:
is there a way to automatically add all ops from `_ops._c10d_functional` and `_ops._c10d_functional_autograd` into this list, instead of listing them out explicitly?

Reply:
Hi, I try to use "list(torch.ops._c10d_functional)" to get all ops, but it returns an empty list. Any suggestions about getting all ops automatically?
```python
>>> import torch
>>> list(torch.ops._c10d_functional)
[]
```

Reply:
Hmm yeah seems there is no clean way to get this list.

Wonder would you like to move this to https://github.com/pytorch/pytorch/blob/e0f1bf14a4fff8f8cbe69c426161beafdbaacf0e/torch/distributed/_functional_collectives.py#L963, by doing something like the following at that file:
```python
    lib_impl.impl("all_to_all_single", _all_to_all_single_meta, "Meta")
    lib_impl.impl("broadcast", _broadcast_meta, "Meta")
    lib_impl.impl("broadcast_", _broadcast__meta, "Meta")
    
    # newly added ops below
    torch.fx.node.has_side_effect(torch._ops._c10d_functional.wait_tensor.default)
    torch.fx.node.has_side_effect(torch_ops._c10d_functional.all_gather_into_tensor_out.default)
    ...
```

Also we likely only need the ones with `.default` overload name, and don't need to add the ones with no overload name.

Thanks!

Reply:
Hi, changed to use has_side_effect in _functional_collectives.py file.

> Also we likely only need the ones with .default overload name, and don't need to add the ones with no overload name.

BTW, could you elaborate more about why we only need the ones with `.default` overload name? I saw the fx graph passed to backend compiler actually **has ops with no overload name**. But those ops are indeed not removed even we don't mark them "has side effect".

Reply:
hmm curious do you have an example of the FX graph that has collective ops with no overload name?

Reply:
A small example:
```python
    def test_keep_collectives_e2e(self):
        import torch.distributed as dist
        from torch._dynamo import compiled_autograd
        from torch._dynamo.backends.common import aot_autograd

        def grad_sync_clear(param, keep_grad):
            dist.all_reduce(param.grad, op=dist.ReduceOp.SUM)
            if not keep_grad:
                param.grad = None

        class TestModule(torch.nn.Module):
            def __init__(self, ioc):
                super().__init__()
                self.fc1 = torch.nn.Linear(ioc, ioc, bias=False)
                self.fc2 = torch.nn.Linear(ioc, ioc, bias=False)

                self.grad_acc_hooks = []
                self.grad_acc = []
                self.params = [self.fc1.weight, self.fc2.weight]
                for i, param in enumerate(self.params):
                    keep_grad = bool(i)

                    def wrapper(param, keep_grad):
                        param_tmp = param.expand_as(param)
                        grad_acc = param_tmp.grad_fn.next_functions[0][0]

                        def grad_acc_hook(*notneeded):
                            grad_sync_clear(param, keep_grad)

                        self.grad_acc.append(grad_acc)
                        self.grad_acc_hooks.append(
                            grad_acc.register_hook(grad_acc_hook))

                    wrapper(param, keep_grad)

            def forward(self, x):
                x = self.fc1(x)
                x = self.fc2(x)
                return x.sum()

        def compiler_fn(gm):

            def inner_compiler(graph_module, example_inputs):
                print(graph_module)

                def fwd_compiler(gm_, ei_):
                    print(gm_)
                    return gm_

                return aot_autograd(fw_compiler=fwd_compiler)(graph_module, example_inputs)

            return torch.compile(gm, backend=inner_compiler, fullgraph=True)

        torch.distributed.init_process_group(
            backend="fake",
            world_size=2,
            rank=0,
            store=FakeStore(),
        )

        bs = 64
        ioc = 128

        model = TestModule(ioc)

        input = torch.randn([bs, ioc])
        loss = model(input)
        with compiled_autograd.enable(compiler_fn):
            loss.backward()
        torch.distributed.destroy_process_group()
```

The graph printed from inner_compiler and fwd_compiler are:
```python
def forward(self, L_inputs_ : list):
    l_inputs_ = L_inputs_
    getitem = l_inputs_[0]
    getitem_1 = l_inputs_[1]
    getitem_2 = l_inputs_[2]
    getitem_4 = l_inputs_[4];  l_inputs_ = None
    expand = torch.ops.aten.expand.default(getitem, [64, 128]);  getitem = None
    t = torch.ops.aten.t.default(expand)
    mm = torch.ops.aten.mm.default(t, getitem_2);  t = getitem_2 = None
    t_1 = torch.ops.aten.t.default(mm);  mm = None
    t_2 = torch.ops.aten.t.default(getitem_1);  getitem_1 = None
    mm_1 = torch.ops.aten.mm.default(expand, t_2);  expand = t_2 = None
    t_3 = torch.ops.aten.t.default(t_1);  t_1 = None
    new_grad = torch.clone(t_3);  t_3 = None
    tensor = torch.ops._c10d_functional.all_reduce(new_grad, 'sum', '0')
    wait_tensor = torch.ops._c10d_functional.wait_tensor(tensor);  tensor = None
    copy_ = new_grad.copy_(wait_tensor);  wait_tensor = copy_ = None
    t_4 = torch.ops.aten.t.default(mm_1);  mm_1 = None
    mm_2 = torch.ops.aten.mm.default(t_4, getitem_4);  t_4 = getitem_4 = None
    t_5 = torch.ops.aten.t.default(mm_2);  mm_2 = None
    t_6 = torch.ops.aten.t.default(t_5);  t_5 = None
    new_grad_1 = torch.clone(t_6);  t_6 = None
    tensor_1 = torch.ops._c10d_functional.all_reduce(new_grad_1, 'sum', '0')
    wait_tensor_1 = torch.ops._c10d_functional.wait_tensor(tensor_1);  tensor_1 = None
    copy__1 = new_grad_1.copy_(wait_tensor_1);  new_grad_1 = wait_tensor_1 = copy__1 = None
    return (new_grad,)

def forward(self, arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1):
    expand = torch.ops.aten.expand.default(arg0_1, [64, 128]);  arg0_1 = None
    t = torch.ops.aten.t.default(expand)
    mm = torch.ops.aten.mm.default(t, arg2_1);  t = arg2_1 = None
    t_1 = torch.ops.aten.t.default(mm);  mm = None
    t_2 = torch.ops.aten.t.default(arg1_1);  arg1_1 = None
    mm_1 = torch.ops.aten.mm.default(expand, t_2);  expand = t_2 = None
    t_3 = torch.ops.aten.t.default(t_1);  t_1 = None
    clone = torch.ops.aten.clone.default(t_3);  t_3 = None
    all_reduce = torch.ops._c10d_functional.all_reduce.default(clone, 'sum', '0')
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_reduce);  all_reduce = None
    copy = torch.ops.aten.copy.default(clone, wait_tensor);  clone = wait_tensor = None
    t_4 = torch.ops.aten.t.default(mm_1);  mm_1 = None
    mm_2 = torch.ops.aten.mm.default(t_4, arg4_1);  t_4 = arg4_1 = None
    t_5 = torch.ops.aten.t.default(mm_2);  mm_2 = None
    t_6 = torch.ops.aten.t.default(t_5);  t_5 = None
    clone_1 = torch.ops.aten.clone.default(t_6);  t_6 = None
    all_reduce_1 = torch.ops._c10d_functional.all_reduce.default(clone_1, 'sum', '0');  clone_1 = None
    wait_tensor_1 = torch.ops._c10d_functional.wait_tensor.default(all_reduce_1);  all_reduce_1 = wait_tensor_1 = None
    return (copy,)
```
If we may do dce inside inner_compiler but before fwd_compiler, then we may need those ops with no overload name? I'm not sure if we will do that.

Reply:
yeah the reason that the collectives in `inner_compiler` is not DCE'ed is because `copy__1` uses `wait_tensor_1` (and I believe for some reason we don't DCE the `.copy_`, so all of them are kept there).

I think to be safe, it makes sense to add the no-overload versions to `torch.fx.node.has_side_effect` as well.

Thanks for looking into this!

Reply:
Okay, let me add

Reply:
added. could you help to trigger the awaiting ci workflow as well as merge job?